app/views/layouts/application.html.erb

```diff
@@ -195,6 +195,22 @@
 </div>
 <script type="text/javascript" nonce="<%= content_security_policy_script_nonce %>">
 
+    // error handling for auth errors when getting user access tokens
+    if (window.SCP.userAccessToken === '' && <%= user_signed_in? %>) {
+        var errorTitle = 'Please sign in again';
+        var errorText = '<p class="text-danger">There appears to be an issue with you session - ' +
+            'please sign out and back in again to continue using Single Cell Portal. </p>';
+        var signOutBtn = "<span class='pull-right'><a class='btn btn-primary' data-method='delete' href='/single_cell/users/sign_out'>" +
+            "<span class='fas fa-sign-out-alt fa-fw'></span> Sign Out</a></span>";
+        var errorMsg = errorText + signOutBtn; // add empty paragraph for padding
+        $('#generic-modal-spinner').html(errorMsg + "<p>&nbsp;</p>");
+        $('#generic-modal-spinner').removeClass('spinner-target');
+        $('#generic-modal-title').html(errorTitle);
+        $('#generic-modal-title').removeClass('text-center');
+        $('#generic-modal-footer').remove();
+        $('#generic-modal').modal('show');
+    }
```

Comment:
I don't consider it blocking, but if it's easy, I'd set `window.SCP.userSignedIn = <%= user_signed_in? %>;` here, move this into a new JavaScript file, then `import` and invoke this code in `packs/application.js`.

That'd better decouple JS and Rails, giving IDE checks for errors, and some front-end testability outside the UI test suite.

Reply:
Yeah, that makes sense to me.  I'll give that a shot - doesn't seem too complicated to do, and I agree it will be better for future maintainability.